### ISSUE TITLE
Create new option to prevent flattened objects

### ISSUE THREAD

**Issue by hayatae** [prettydiff/prettydiff#201]
It looks like prettydiff is trying to put all properties of an object onto a single line instead of breaking them up one per line. Tested with 1.15.12.

Input:
```javascript
module.exports = React.createClass({
  displayName: 'Test',
  getInitialState: function () {
    return {
      a: false,
      b: true,
      c: false,
      d: true
    };
  },
  render: function (){
      return <div/>;
  }
});
```

Output:
```javascript
module.exports = React.createClass({
    displayName: 'Test', getInitialState: function () {
        return {a: false, b: true, c: false, d: true};
    },
    render: function () {
        return <div/>;
    }
});
```

Expected:
Given the input above, the line breaks shouldn't change at all. It doesn't make any sense why displayName and getInitialState are being put on the same line, but the render property is being left alone. Also, all properties inside the return object are being moved onto a single line. This is often not desirable as it makes the code very difficult to read.

I didn't see any options for changing how objects are broken up or not broken up.

**Comment by prettydiff** [comment]
I made a change to the logic I used for destructuring.  It was pretty complex, and thus subject to failure from edge cases.  Now the logic is uber simple.  Here are the rules:

Objects appear on a single line of code, but only if:

* flat, as in not containing functions, arrays, objects, or other non-primitives
* preceeded by either `return` or directly wrapped in parens (as in a method or expression)

As I was writing this I did notice an example that I missed in testing:  That when an object is contained in parens it should be the only thing in parens, so if it is part of a comma separated list it should be broken onto multiple lines.  I will update the code to account for this scenario.

I could, though, create an option so that objects are never flattened onto a single line.

**Comment by hayatae** [comment]
Yeah, we'll definitely need an option if this behavior is intentional and isn't going to change. Otherwise I'll still have the above problem where displayName and the first method in my react class object are being put on the same line. According to the rules, since createClass only takes a single object, it's always going to try to put all properties on the same line.

Of course this is going to require another update to atom-beautify as well since I don't think there's currently an option at all for controlling how objects are condensed or expanded.

**Comment by prettydiff** [comment]
`getInitialState` should have been indented.  I did not catch this and that is a bug.  I have since fixed this with v1.15.13.  Please let me know if this is more acceptable.  I am still considering creating an option to prevent object/array flattening.

**Comment by hayatae** [comment]
Aaah, thanks! That's looking much better!

Options are always good in my book. Thanks for the consideration and the fix. :)

**Comment by hayatae** [comment]
I believe there might be an additional bug in the current code. The following object is being flattened into a single line:

```
return {
        someVarA: someOtherObj.someVarA,
        someVarB: someOtherObj.someVarB,
        someVarC: someOtherObj.someVarC,
        someVarD: someOtherObj.someVarD,
        someVarE: someOtherObj.someVarE
      };
```

Shouldn't objects not be flattened if their properties are object references? Or is this intended behavior? Again, flattening a large object like this creates a very unreadable line of code. The only time flattening even makes sense to me is when there is only a single primitive property within an object. Then again, maybe this will just be addressed by the option?

**Comment by prettydiff** [comment]
Good catch!  I will certainly update this per your suggestion and let you know once I have this resolved.

**Comment by hayatae** [comment]
Excellent. Thanks! I'll let you know if I run into any additional problems.

**Comment by animecyc** [comment]
Was curious to see if there was any headway on this -- can we expect to see something in the coming days/weeks?

**Comment by prettydiff** [comment]
@animecyc Sorry for not communicating my timeline.  I will release this when I release https://github.com/prettydiff/prettydiff/issues/202.  I had already started work on the TypeScript support shortly after I resolved the original defect described here.  I hope to have this released sooner than later, but TypeScript is proving to be very challenging to support from a JavaScript parser as it is far more similar to Java/C# than JavaScript.

It is a reasonable bet that once I nail TypeScript support I shouldn't (in theory) have much work to do to support Java or C# in a later unplanned enhancement.

I had no idea there was so much interest in TypeScript, but clearly there is: https://github.com/Glavin001/atom-beautify/issues/601
